lib/src/generate/build.dart

```diff
@@ -0,0 +1,153 @@
+  if (!pubspec.existsSync()) {
+    throw 'Build scripts must be invoked from the top level package directory, '
+        'which must contain a pubspec.yaml';
+  }
+  var yaml = loadYaml(pubspec.readAsStringSync()) as YamlMap;
+  if (yaml['name'] == null) {
+    throw 'You must have a `name` specified in your pubspec.yaml file.';
+  }
+  return yaml['name'];
+}();
+
+/// Validates the phases.
+void _validatePhases(List<List<Phase>> phaseGroups) {
+  if (phaseGroups.length > 1) {
+    // Don't support using generated files as inputs yet, so we only support
+    // one phase.
+    throw new UnimplementedError(
+        'Only one phase group is currently supported.');
+  }
+}
+
+/// Validates the phases, for now that just means making sure they are all only
+/// running on the current package.
+///
+/// Returns a [Future<BuildResult>] which completes once all [Phase]s are done
+/// running.
+Future<BuildResult> _runPhases(List<List<Phase>> phaseGroups) async {
+  var outputs = [];
+  for (var group in phaseGroups) {
+    for (var phase in group) {
+      var inputs = _assetIdsFor(phase.inputSets);
+      for (var builder in phase.builders) {
+        // TODO(jakemac): Optimize, we can run all the builders in a phase
+        // at the same time instead of sequentially.
+        await _runBuilder(builder, inputs, outputs);
+      }
+    }
+  }
+  return new BuildResult(BuildStatus.Success, BuildType.Full, outputs);
+}
+
+/// Gets all [AssetId]s matching [inputSets] in the current package.
+List<AssetId> _assetIdsFor(List<InputSet> inputSets) {
+  var ids = [];
+  for (var inputSet in inputSets) {
+    var files = _filesMatching(inputSet);
+    for (var file in files) {
+      var segments = file.uri.pathSegments;
+      var newPath = path.joinAll(segments.getRange(
+          segments.indexOf(inputSet.package) + 1, segments.length));
+      ids.add(new AssetId(inputSet.package, newPath));
+    }
+  }
+  return ids;
+}
+
+/// Returns all files matching [inputSet].
+Set<File> _filesMatching(InputSet inputSet) {
+  if (inputSet.package != _localPackageName) {
+    throw new UnimplementedError('Running on packages other than the '
+        'local package is not yet supported');
+  }
+
+  var files = new Set<File>();
+  for (var pattern in inputSet.filePatterns) {
+    files.addAll(new Glob(pattern).listSync(followLinks: false).where(
+        (e) => e is File && !_ignoredDirs.contains(path.split(e.path)[1])));
+  }
+  return files;
+}
+
+/// Runs [builder] with [inputs] as inputs.
+Future _runBuilder(
+    Builder builder, List<AssetId> inputs, List<Asset> outputs) async {
+  for (var input in inputs) {
+    var expectedOutputs = builder.declareOutputs(input);
+    var inputAsset = new Asset(input, await _reader.readAsString(input));
+    var buildStep =
+        new BuildStep(inputAsset, expectedOutputs, _reader, _writer);
+    await builder.build(buildStep);
+    await buildStep.outputsCompleted;
+    outputs.addAll(buildStep.outputs);
```

Comment:
Agreed that using a return value is a simpler pattern in this scenario (and in general). Sometimes my c++ days come back to haunt me.